### ISSUE TITLE
[Bug]: I cannot install the module template

### ISSUE THREAD

**Issue by bmatyasb** [jasp-stats/jasp-issues#3083]
### JASP Version

0.19.2

### JASP Module

Unrelated

### What OS are you seeing the problem on?

Windows 11

### Bug Description

I would like to add extra functionality to JASP for myself. I think I have gone through the tutorial on adding modules. I saw that there is a template module. I downloaded it and thought I'd try loading it into JASP, testing to see if I had the simplest dependencies. But unfortunately it didn't work. I got an error message, I'll screenshot it. Is it just me, or is the template called jaspModule not really installable "by itself", just as downloaded?
![Clipboard_12-12-2024_01](https://github.com/user-attachments/assets/f326437f-e4b1-4f31-9ba5-687c584f19cd)

### Expected Behaviour

Actually, I don't know : )

### Steps to Reproduce

Download the template as a zip file (https://github.com/jasp-stats/jaspModuleTemplate).
Unpack, repack as targz.
JASP -> developer mode -> top right corner -> select the targz. 

(in fact I tried to specify the source folder in the options (obviously not the targz file), and still no luck. )

### Final Checklist

- [X] I have included a screenshot showcasing the issue, if possible.
- [ ] I have included a JASP file (zipped) or data file that causes the crash/bug, if applicable.
- [X] I have accurately described the bug, and steps to reproduce it.

**Comment by JorisGoosen** [comment]
@bmatyasb yes there is some crucial parts of the template missing that would allow it to just be installed like that.

Currently @PabRod is working hard on upgrading the documentation and making the template immediately installable. 
He is pretty far with this.

**Comment by RensDofferhoff** [comment]
You could build the module using RStudio and than load it from the resulting libpath.
This workflow might be easier as you are likely already familiar with using RStudio.

**Comment by bmatyasb** [comment]
thanks! I am very much looking forward to it. i would like to add tidyplot
as a JASP module to allow a truly customized plot creation for anyone who
wants it

**Comment by tomtomme** [comment]
This would be amazing!!!

**Comment by JorisGoosen** [comment]
> thanks! I am very much looking forward to it. i would like to add tidyplot as a JASP module to allow a truly customized plot creation for anyone who wants it

This libpaths feature is included in the latest release that is currently live.
So you can give it a go if you like.

**Comment by tomtomme** [comment]
@boutinb 
I guess the "installing" label is not the best here - could we get a "developer" tag or soemthing like that?

**Comment by JorisGoosen** [comment]
@bmatyasb we have a mattermost server where we do most of our internal communication. There is also a channel there specifically to help people who want to develop modules.

Would you like to be added to this?
Perhaps you can then send an email to info@jasp-stats.org, then we know what email to add. (Github will probably redact it here and you might not want to post it online anyway)

**Comment by bmatyasb** [comment]
thank you very much for your trust! :D For now I'd like to try it out to
see if it's not too much for me. I have only developed Shiny apps so far,
and I don't want to just promise. I'll report back soon as I manage to
squeeze out some basic tidyplot functionality.

**Comment by boutinb** [comment]
> @boutinb
> I guess the "installing" label is not the best here - could we get a "developer" tag or soemthing like that?

Done

**Comment by bmatyasb** [comment]
Hi guys,
I think I can integrate the whole tidyplot into JASP, the basics are already working, it's not hard to add layers from here. I hope to have a version ready for testing by the end of January. For now it's in the Descriptives menu.
![image](https://github.com/user-attachments/assets/e1e0f296-357c-4a2a-939b-62f81e43a6aa)

**Comment by tomtomme** [comment]
@bmatyasb 
Wow, just wow! Very nice to see. This will be a very valuable addition. Thank you so much!

**Comment by boutinb** [comment]
Impressive @bmatyasb !

**Comment by bmatyasb** [comment]
I'm doing pretty well. I think I'll be done by mid-January. The tidyplotting is really quite easy, very customizable :) However, I have a question. When I'm done, should I just compress the Package I created in RStudio and send it to you? I suppose you then build -> install package in RStudio in the same way as I did, and then load it as a developer module in JASP. 
(...Unfortunately, I have not yet managed to compress it into tar.gz and install it as a plain module).

To which e-mail will I be able to send it?

![image](https://github.com/user-attachments/assets/fd0cbff0-bb69-434c-82e2-916f856e0794)

**Comment by tomtomme** [comment]
@bmatyasb 
very very cool.

As far as I know, how to submit your module as tar.gz is described here in step 3, but this may be outdated: 
https://github.com/jasp-stats/jasp-desktop/blob/development/Docs/development/jasp-module-workflow.md

@boutinb 
can you confirm?

**Comment by shun2wang** [comment]
> I'm doing pretty well. I think I'll be done by mid-January. The tidyplotting is really quite easy, very customizable :) However, I have a question. When I'm done, should I just compress the Package I created in RStudio and send it to you? I suppose you then build -> install package in RStudio in the same way as I did, and then load it as a developer module in JASP. 

Great works!

you can just create a github repo and we will fork your repo to build into JASP.

**Comment by EJWagenmakers** [comment]
Let me just chime in and echo how cool this is!!
EJ

**Comment by bmatyasb** [comment]
dear developers! I have a very important question.
I would like to have an option to edit already created plots to a plot grid
using cowplot(). But i need to have the plots created with the plot builder
function somehow saved in a "global environment", from where i can call
them back by name.
Currently what happens is that if I give a plot a title, it is saved with
that name, but I don't know where. I guess the usual global env won't work
here.
The whole end result would be that in a plot grid option, I would type in a
text field the names of the plots I have created so far with the plot
builder, and they would be put in a nice order. but to do that, I would
have to get them in the subpane of their names.

I'll show you a snippet to see if it's clear where I want to save and what
I want to save.

   # 8. Attach the plot to jaspResults
    jaspResults[["tidyPlot"]] <- jaspPlot

    # 9. If we have a non-empty title, save the plot to the .GlobalEnv
    if (!is.null(plotTitle) && plotTitle != "") {
      assign(plotTitle, jaspPlot$plotObject, envir = .GlobalEnv)

      # Also store the title in the known titles
      currentTitles <- jaspResults[["titleList"]]$object
      if (!plotTitle %in% currentTitles) {
        currentTitles <- c(currentTitles, plotTitle)
      }
      jaspResults[["titleList"]]$object <- currentTitles
    }
  }

  # 10. Expose list of known plot titles to QML
  allTitles <- jaspResults[["titleList"]]$object
  titlesForQml <- paste(allTitles, collapse = "\n")
  createJaspQmlSource("savedPlotTitlesQML", titlesForQml)

  return(jaspResults)

**Comment by bmatyasb** [comment]
or should everything be in japs results, because that's the "global env" in
jasp? :D sorry if I couldn't make the problem clear.

**Comment by bmatyasb** [comment]
Dear Developers,
making plots one by one is only one thing. It would be nice if you could
create plot layouts.
I can do that with another function. So inside the module there would be a
Plot builder and a Plot grid.
Except that in the Plot grid I would have to somehow access the plots
created in the Plot builder. The obvious thing to do would be to give each
plot a name, which I could call in the Plot grid. I have tried a lot, but I
don't know how to do it.
How can I leave the plot created in the Plot builder "in the JASP
environment" and call the same plot in another function of the module. Is
there a way to do this? I have tried with createJaspState, but no success.

**Comment by JorisGoosen** [comment]
@bmatyasb how will you allow users to add multiple plots?

As in, they would have to come from the same analysis. So this would need some sort of list or entryfield for having multiple plots right?

The relevant option can then be used as dependencies for a jaspContainer to keep the plots in?
Because storing the plotobjects in the state seems a bit impractical and would probably become quite a hassle.

Perhaps joining our chatserver could be practical? The offer still stands and it would allow us to communicate easier.

**Comment by shun2wang** [comment]
>Perhaps joining our chatserver could be practical? The offer still stands and it would allow us to communicate easier.

I had just invited @bmatyasb to join in our channel ^^ please check your e-mail messages.

**Comment by tomtomme** [comment]
@bmatyasb 
how is it going? Anything you need?

**Comment by bmatyasb** [comment]
I'm getting there. thanks! i think i'm fine without help! I'm really trying
to finish this month :)

**Comment by tomtomme** [comment]
> [@bmatyasb](https://github.com/bmatyasb) yes there is some crucial parts of the template missing that would allow it to just be installed like that.
> 
> Currently [@PabRod](https://github.com/PabRod) is working hard on upgrading the documentation and making the template immediately installable. He is pretty far with this.

@PabRod 
any news on the "making the template immediately installable" part?

**Comment by PabRod** [comment]
Not yet, @tomtomme. I've been taking care of some urgent issues lately, but you'd like to know I'll be back to this starting tomorrow.

**Comment by PabRod** [comment]
Update: we are close to [merging](https://github.com/jasp-stats/jaspModuleTemplate/pull/12).
